nomad/data_source_dynamic_host_volume.go

```diff
@@ -0,0 +1,205 @@
+				Computed:    true,
+			},
+			"node_pool": {
+				Description: "Node pool",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			"parameters": {
+				Description: "Parameters",
+				Type:        schema.TypeMap,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Computed: true,
+			},
+			"plugin_id": {
+				Description: "Plugin ID",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			"state": {
+				Description: "State",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			"capacity_bytes": {
+				Computed: true,
+				Type:     schema.TypeInt,
+			},
+			"capacity_min_bytes": {
+				Computed: true,
+				Type:     schema.TypeInt,
+			},
+			"capacity_max_bytes": {
+				Computed: true,
+				Type:     schema.TypeInt,
+			},
+		},
+	}
+}
+
+// dynamicHostVolumeRead gets a dynamic host volume from Nomad. This same
+// function is used for the data source and the create/register resources
+func dynamicHostVolumeRead(d *schema.ResourceData, meta any) error {
+	client := meta.(ProviderConfig).client
+
+	ns, id := getDynamicHostVolumeNamespacedID(d)
+	vol, _, err := client.HostVolumes().Get(id, &api.QueryOptions{Namespace: ns})
+	if err != nil {
+		return fmt.Errorf("Failed to get information about %q: %v", id, err)
+	}
+
+	sw := helper.NewStateWriter(d)
+	mapping := map[string]any{
+		"capability":         flattenDynamicHostVolumeCapabilities(vol.RequestedCapabilities),
+		"capacity_bytes":     vol.CapacityBytes,
+		"capacity_max_bytes": vol.RequestedCapacityMaxBytes,
+		"capacity_min_bytes": vol.RequestedCapacityMinBytes,
+		"constraint":         flattenConstraints(vol.Constraints),
+		"host_path":          vol.HostPath,
+		"id":                 vol.ID,
+		"name":               vol.Name,
+		"namespace":          vol.Namespace,
+		"node_id":            vol.NodeID,
+		"node_pool":          vol.NodePool,
+		"parameters":         vol.Parameters,
+		"plugin_id":          vol.PluginID,
+		"state":              vol.State,
+	}
+	for k, v := range mapping {
+		sw.Set(k, v)
+	}
+
+	// TODO: these fields may have been set by the config but if not, we want to
+	// get them from the API. But right now these are getting changed anyways,
+	// so we detect mutations from ex. "1GiB" => "1.0 GiB"
+	if d.Get("capacity") == "" && vol.CapacityBytes > 0 {
+		sw.Set("capacity", humanize.IBytes(uint64(vol.CapacityBytes)))
+	}
+	if d.Get("capacity_min") == "" && vol.RequestedCapacityMinBytes > 0 {
+		sw.Set("capacity_min", humanize.IBytes(uint64(vol.RequestedCapacityMinBytes)))
+	}
+	if d.Get("capacity_max") == "" && vol.RequestedCapacityMaxBytes > 0 {
+		sw.Set("capacity_max", humanize.IBytes(uint64(vol.RequestedCapacityMaxBytes)))
+	}
+
+	d.SetId(id)
+	return nil
+}
+
+// getDynamicHostVolumeNamespacedID returns the namespace and ID for a dynamic
+// host volume from the resource data
+func getDynamicHostVolumeNamespacedID(d *schema.ResourceData) (string, string) {
+	id := d.Id()
```

Comment:
this causes an error in the `data` source, because it's empty until there's state?

```hcl
resource "nomad_dynamic_host_volume" "one" {
  name      = "volume-one"
  plugin_id = "mkdir"
  namespace = "default"
  capability {
    access_mode     = "single-node-writer"
    attachment_mode = "file-system"
  }
}

data "nomad_dynamic_host_volume" "one" {
  id = nomad_dynamic_host_volume.one.id
}
```

> Error: Failed to get information about "": Unexpected response code: 404 (resource not found)

if I do this, then it fixes it, but I haven't checked to see if there are any other implications of doing so

```suggestion
func getDynamicHostVolumeNamespacedID(d *schema.ResourceData) (string, string) {
	id := d.Get("id").(string)
```

Reply:
This got refactored a bunch and I don't think I re-tested the data source. Lemme give that a try and report back.

Reply:
Ok, checked the current build end-to-end and that seems to work.